File: DVSM/commercial/binframe_adapter.py
# SPDX-License-Identifier: AGPL-3.0-only
# Commercial use beyond AGPL-3.0 requires a separate license — see LICENSE-COMMERCIAL.md.
"""
binframe_adapter.py — ingest REAL DVSM kernel telemetry dumps (the B3 "lift"): turn reference-relative
results into KERNEL-relative ones by reading the bytes the shipped Rust kernel actually emits.

The DVSM kernel emits ABI-stable `repr(C)` frames over its C ABI (`dvsm_step` fills a frame struct):
  • SCHEMA_TELEM  (dvsm_one_file.rs `BinaryFrame`): rich diagnostics — energy, novelty, stress, stiffness,
                  omega_norm, entropy, drift, resonance_peak, ghost, contained, emitted. NO replay hash.
  • SCHEMA_ABI    (dvsm_v20 `run_profile` record): id, energy, stress, entropy, ghost, contained, hash.
                  Carries the FNV-1a replay hash; fewer diagnostic channels.

This adapter (1) parses either layout with a record-size VALIDATION that flags a layout/endianness mismatch
instead of emitting silent garbage (`parsed ≠ correct`), (2) LIFTS the obligations the emitted telemetry can
actually support to kernel-relative graded `ObligationResult`s, and (3) **explicitly declares which obligations
cannot be lifted from frame dumps alone** — the forbidden-coupling air-gap checks (Ω→V, ν→λ) need the velocity
`V` and dissipation `λ`, which the public frame does NOT emit. `emitted-telemetry ≠ full-state`.

LAYOUT CAVEAT (honest): the default formats assume packed little-endian. `repr(C)` may insert trailing padding
on your target (e.g. sizeof rounded up to the u64 alignment). The adapter validates `len(body) % rec_size == 0`
and surfaces a `layout_mismatch` — verify the format against your build's `sizeof(BinaryFrame)` before trusting
parsed values. Customers can pass their own `BinFrameSchema(fmt=…, fields=…)`.
"""
from __future__ import annotations

import math
import os
import struct
import sys
import tempfile
from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

_HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _HERE)
sys.path.insert(0, os.path.join(_HERE, ".."))
sys.path.insert(0, os.path.join(_HERE, "..", "..", "weltwerk", "verify"))
from invariant_ledger import ObligationResult, CLOSED, BOUNDED, VIOLATED, UNDERDETERMINED  # noqa: E402

GHOST_NAMES = ("Nominal", "Collapse", "Diffuse", "Echo", "Burst", "Trap", "Vacuum", "Denatured")
U_MAX_DEFAULT = 100.0


@dataclass(frozen=True)
class BinFrameSchema:
    """A `struct` format + field names for one on-disk frame record. `fmt` uses struct syntax; the leading
    '<' selects packed little-endian (verify against your build's actual sizeof — see LAYOUT CAVEAT)."""
    name: str
    fmt: str
    fields: Tuple[str, ...]

    @property
    def rec_size(self) -> int:
        return struct.calcsize(self.fmt)


# dvsm_one_file.rs BinaryFrame — rich diagnostics, packed little-endian assumption.
SCHEMA_TELEM = BinFrameSchema(
    "dvsm_one_file.BinaryFrame",
    "<QffffffffBBBB",
    ("frame", "energy", "novelty", "stress", "stiffness", "omega_norm", "entropy", "drift",
     "resonance_peak", "ghost", "contained", "emitted", "_pad"))

# dvsm_v20 run_profile record — carries the replay hash.
SCHEMA_ABI = BinFrameSchema(
    "dvsm_v20.run_profile",
    "<QdddBBQ",
    ("frame", "energy", "stress", "entropy", "ghost", "contained", "hash"))


@dataclass(frozen=True)
class ParseReport:
    schema: str
    n_records: int
    rec_size: int
    leftover_bytes: int
    layout_mismatch: bool
    nonfinite: int           # rows with a non-finite float (a parse/endianness ghost)

    def ok(self) -> bool:
        return not self.layout_mismatch and self.nonfinite == 0 and self.n_records > 0


def read_binframes(path: str, schema: BinFrameSchema = SCHEMA_TELEM,
                   header_lines: int = 0) -> Tuple[List[dict], ParseReport]:
    """Parse a binary dump into dict rows + a ParseReport. `header_lines` skips that many leading text lines
    (e.g. the run_profile header). A non-zero leftover or a non-finite float is surfaced, not hidden."""
    with open(path, "rb") as f:
        data = f.read()
    return parse_binframes(data, schema, header_lines)


def parse_binframes(data: bytes, schema: BinFrameSchema = SCHEMA_TELEM,
                    header_lines: int = 0) -> Tuple[List[dict], ParseReport]:
    off = 0
    for _ in range(header_lines):
        nl = data.find(b"\n", off)
        if nl < 0:
            break
        off = nl + 1
    body = data[off:]
    rec = schema.rec_size
    n, rem = divmod(len(body), rec)
    rows: List[dict] = []
    nonfinite = 0
    for i in range(n):
        vals = struct.unpack_from(schema.fmt, body, i * rec)
        row = {f: v for f, v in zip(schema.fields, vals) if not f.startswith("_")}
        if any(isinstance(v, float) and not math.isfinite(v) for v in row.values()):
            nonfinite += 1
        rows.append(row)
    report = ParseReport(schema.name, n, rec, rem, rem != 0, nonfinite)
    return rows, report


def write_binframes(path: str, rows: List[dict], schema: BinFrameSchema = SCHEMA_TELEM,
                    header: bytes = b"") -> None:
    """Pack rows back to the on-disk layout (for round-trip tests and fixtures)."""
    with open(path, "wb") as f:
        if header:
            f.write(header)
        for r in rows:
            f.write(struct.pack(schema.fmt, *[r.get(fld, 0) for fld in schema.fields]))


def with_next(rows: List[dict], fields: Tuple[str, ...]) -> List[dict]:
    """Add `<field>_next` columns (the next frame's value) so a per-frame dump can feed a diagnostic→dynamics
    probe. Drops the final row (no successor). The kernel emits per-frame; the audit needs t vs t+1."""
    out = []
    for a, b in zip(rows, rows[1:]):
        r = dict(a)
        for fld in fields:
            r[fld + "_next"] = b[fld]
        out.append(r)
    return out


def ghost_census(rows: List[dict]) -> Dict[str, int]:
    out = {g: 0 for g in GHOST_NAMES}
    for r in rows:
        g = int(r.get("ghost", 0))
        out[GHOST_NAMES[g] if 0 <= g < len(GHOST_NAMES) else "Nominal"] += 1
    return out


# ---- lift: obligations the emitted telemetry CAN support, kernel-relative -------------------------
def containment(rows: List[dict], u_max: float = U_MAX_DEFAULT, field: str = "energy") -> ObligationResult:
    mx = max((float(r[field]) for r in rows if field in r), default=float("nan"))
    status = BOUNDED if math.isfinite(mx) and mx < u_max else VIOLATED
    return ObligationResult(
        "DVSM-3-kernel", "‖Z‖ stays under U_MAX on the REAL emitted telemetry", status,
        f"max(energy)={mx:.3f} over {len(rows)} real frames (bound={u_max})",
        "boundedness for all inputs — only this dumped run; empirical-boundedness ≠ certified",
        "a real dump whose energy reaches U_MAX without GhostSnap recovery")


def replay_parity(rows_a: List[dict], rows_b: List[dict], field: str = "hash") -> ObligationResult:
    a = [r.get(field) for r in rows_a]
    b = [r.get(field) for r in rows_b]
    status = CLOSED if a and a == b else VIOLATED
    return ObligationResult(
        "DVSM-6-kernel", "two real dumps from the same seed share an identical replay-hash sequence", status,
        f"{len(a)} vs {len(b)} frames; hash sequences identical = {a == b}",
        "CORRECTNESS or cross-precision parity — integrity ≠ truth; hash ≠ reality",
        "identical seed yielding divergent emitted hashes")


# obligations that CANNOT be lifted from a public frame dump (and why) — the honest B3 boundary
_NEEDS = {
    "DVSM-7 (Ω→V air-gap)": ("v", "velocity V is not emitted in the public frame; the Ω→V air-gap needs it"),
    "DVSM-4 (ν→λ air-gap)": ("lambda_eff", "dissipation λ is not emitted in the public frame; the ν→λ air-gap needs it"),
}


def non_liftable(rows: List[dict]) -> List[Tuple[str, str]]:
    """List forbidden-coupling obligations that the emitted fields cannot support, with the reason. These
    require richer instrumentation (a custom telemetry build), not a verdict from the public frame."""
    have = set(rows[0].keys()) if rows else set()
    return [(oid, reason) for oid, (need, reason) in _NEEDS.items() if need not in have]


def lift(rows: List[dict], schema: BinFrameSchema, *, u_max: float = U_MAX_DEFAULT,
         rows_b: Optional[List[dict]] = None) -> Tuple[List[ObligationResult], List[Tuple[str, str]]]:
    """Lift every obligation the dump supports to kernel-relative, and return the non-liftable ones with why."""
    obs: List[ObligationResult] = []
    if rows and "energy" in rows[0]:
        obs.append(containment(rows, u_max))
    if rows and "hash" in rows[0] and rows_b is not None:
        obs.append(replay_parity(rows, rows_b))
    elif rows and "hash" not in rows[0]:
        obs.append(ObligationResult(
            "DVSM-6-kernel", "replay-hash parity on the real dump", UNDERDETERMINED,
            f"the {schema.name} frame carries no hash field — cannot check replay parity from this schema",
            "anything about reproducibility from a hashless dump",
            "switch to the ABI/run_profile dump (carries the FNV-1a hash)"))
    return obs, non_liftable(rows)


def _safe_remove(path: str) -> None:
    try:
        os.remove(path)
    except OSError:
        pass


def main():
    print("binframe_adapter.py — real DVSM BinaryFrame ingest (the B3 lift)\n")
    # synthesize a real-format dump from the reference (we cannot run the Rust kernel here)
    from dvsm_reference import gen_clean
    trace = gen_clean(3000, seed=1)
    abi_rows = [{"frame": r.t, "energy": r.energy, "stress": r.stress, "entropy": r.entropy,
                 "ghost": GHOST_NAMES.index(r.ghost) if r.ghost in GHOST_NAMES else 0,
                 "contained": r.contained, "hash": r.hash} for r in trace]
    path = os.path.join(tempfile.gettempdir(), "_dvsm_demo_abi.bin")
    write_binframes(path, abi_rows, SCHEMA_ABI, header=b"DVSM-V20 demo R=4\n")
    rows, report = read_binframes(path, SCHEMA_ABI, header_lines=1)
    print(f"  parsed {report.n_records} ABI frames (rec={report.rec_size}B) layout_ok={report.ok()}")
    obs, notliftable = lift(rows, SCHEMA_ABI, rows_b=rows)   # rows_b=rows ⇒ trivially-identical replay
    for o in obs:
        print(f"    [{o.status:9s}] {o.id}: {o.witness}")
    print("  NOT liftable from this dump:")
    for oid, why in notliftable:
        print(f"    - {oid}: {why}")
    print(f"  ghost census: {ghost_census(rows)}")
    _safe_remove(path)
    print("\n  emitted-telemetry ≠ full-state; parsed ≠ correct; integrity ≠ truth.")


if __name__ == "__main__":
    main()

File: Rust/src/binframe_adapter.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! binframe_adapter (Sub-Slice 1A) — the BinaryFrame PARSER, ported from `DVSM/commercial/binframe_adapter.py`.
//! Gateway layer 1, ingestion. This slice is the parser ONLY: read a fixed-record byte stream, validate the
//! layout, flag anomalies into a [`ParseReport`], and return typed rows.
//!
//! Two on-disk schemas (packed little-endian, `repr(C)`):
//!   * [`SCHEMA_TELEM`] (`dvsm_one_file.BinaryFrame`, 44 B) — rich diagnostics: frame + 8 f32 channels
//!     (energy, novelty, stress, stiffness, omega_norm, entropy, drift, resonance_peak) + 4 u8
//!     (ghost, contained, emitted, _pad). NO replay hash.
//!   * [`SCHEMA_ABI`] (`dvsm_v20.run_profile`, 42 B) — frame + 3 f64 (energy, stress, entropy) + 2 u8
//!     (ghost, contained) + u64 hash (the FNV-1a replay hash).
//!
//! ## Faithful to the reference — what this DOES / DOES NOT do
//! - **`lift()` (Sub-Slice 1B, now wired).** Grades the obligations a dump CAN support (containment,
//!   replay-parity) over the ported [`crate::invariant_ledger`], and HONESTLY declares the non-liftable
//!   air-gap checks (Ω→V, ν→λ) because the public frame carries neither `v` nor `lambda_eff`.
//!   `emitted-telemetry ≠ full-state`; `undetected ≠ absent`.
//! - **No "ForbiddenSetViolation."** The Python `ParseReport` has exactly two anomalies — `layout_mismatch`
//!   (a non-zero record remainder) and `nonfinite` (NaN/Inf in a float field). There is no forbidden-flag check
//!   in the reference, so none is invented here. `claim ≠ code`.
//! - **Not zero-copy.** Fields are decoded into typed values (a cheap copy), exactly like the Python rows.
//!
//! ## `does_not_show`
//! `parsed ≠ correct`: the formats assume packed little-endian. A build whose `repr(C)` inserts padding has a
//! different `rec_size`; the report's `layout_mismatch` catches a non-dividing stream, but a coincidentally
//! dividing-but-misaligned stream would parse to garbage. Verify the schema against your build's
//! `sizeof(BinaryFrame)` before trusting values. `emitted-telemetry ≠ full-state` (V and λ are not emitted, so
//! the Ω→V / ν→λ air-gap obligations cannot be lifted from a public frame — surfaced in 1B, not here).

use crate::invariant_ledger::{ObligationResult, ObligationStatus};
use std::io::{BufReader, Read};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FieldType {
    U64,
    F32,
    F64,
    U8,
    Pad, // a byte consumed but dropped from output (mirrors the Python `_`-prefixed field)
}

/// A decoded field value. `PartialEq` only (floats are not `Eq`).
#[derive(Debug, Clone, PartialEq)]
pub enum Field {
    U64(u64),
    F32(f32),
    F64(f64),
    U8(u8),
}

impl Field {
    pub fn as_f64(&self) -> Option<f64> {
        match self {
            Field::F32(x) => Some(*x as f64),
            Field::F64(x) => Some(*x),
            _ => None,
        }
    }
    pub fn as_u64(&self) -> Option<u64> {
        match self {
            Field::U64(x) => Some(*x),
            _ => None,
        }
    }
    pub fn as_u8(&self) -> Option<u8> {
        match self {
            Field::U8(x) => Some(*x),
            _ => None,
        }
    }
    pub fn is_nonfinite(&self) -> bool {
        match self {
            Field::F32(x) => !x.is_finite(),
            Field::F64(x) => !x.is_finite(),
            _ => false,
        }
    }
}

/// A `struct`-format schema: an ordered list of `(field name, type)`.
pub struct Schema {
    pub name: &'static str,
    pub fields: &'static [(&'static str, FieldType)],
}

impl Schema {
    pub fn rec_size(&self) -> usize {
        self.fields
            .iter()
            .map(|(_, ft)| match ft {
                FieldType::U64 | FieldType::F64 => 8,
                FieldType::F32 => 4,
                FieldType::U8 | FieldType::Pad => 1,
            })
            .sum()
    }
}

pub const SCHEMA_TELEM: Schema = Schema {
    name: "dvsm_one_file.BinaryFrame",
    fields: &[
        ("frame", FieldType::U64),
        ("energy", FieldType::F32),
        ("novelty", FieldType::F32),
        ("stress", FieldType::F32),
        ("stiffness", FieldType::F32),
        ("omega_norm", FieldType::F32),
        ("entropy", FieldType::F32),
        ("drift", FieldType::F32),
        ("resonance_peak", FieldType::F32),
        ("ghost", FieldType::U8),
        ("contained", FieldType::U8),
        ("emitted", FieldType::U8),
        ("_pad", FieldType::Pad),
    ],
};

pub const SCHEMA_ABI: Schema = Schema {
    name: "dvsm_v20.run_profile",
    fields: &[
        ("frame", FieldType::U64),
        ("energy", FieldType::F64),
        ("stress", FieldType::F64),
        ("entropy", FieldType::F64),
        ("ghost", FieldType::U8),
        ("contained", FieldType::U8),
        ("hash", FieldType::U64),
    ],
};

/// One parsed row: `(field name, value)`, with `Pad` fields dropped (mirrors the Python `_`-prefix skip).
pub type Row = Vec<(&'static str, Field)>;

/// Look up a field in a row by name.
pub fn field<'a>(row: &'a Row, name: &str) -> Option<&'a Field> {
    row.iter().find(|(n, _)| *n == name).map(|(_, v)| v)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseReport {
    pub schema: &'static str,
    pub n_records: usize,
    pub rec_size: usize,
    pub leftover_bytes: usize,
    pub layout_mismatch: bool,
    pub nonfinite: usize,
}

impl ParseReport {
    pub fn ok(&self) -> bool {
        !self.layout_mismatch && self.nonfinite == 0 && self.n_records > 0
    }
}

/// Parse a fixed-record byte stream into typed rows + a [`ParseReport`]. `header_lines` skips that many leading
/// newline-terminated text lines (e.g. the `run_profile` header) before the binary body. A non-zero remainder
/// (`layout_mismatch`) or any non-finite float (`nonfinite`) is surfaced, never hidden. `parsed ≠ correct`.
pub fn parse_frames(data: &[u8], schema: &Schema, header_lines: usize) -> (Vec<Row>, ParseReport) {
    // skip header text lines
    let mut start = 0usize;
    for _ in 0..header_lines {
        match data[start..].iter().position(|&b| b == b'\n') {
            Some(p) => start += p + 1,
            None => break,
        }
    }
    let body = &data[start..];

    let rec = schema.rec_size();
    let (n, rem) = if rec == 0 { (0, body.len()) } else { (body.len() / rec, body.len() % rec) };

    let mut rows: Vec<Row> = Vec::with_capacity(n);
    let mut nonfinite = 0usize;
    for i in 0..n {
        let (row, row_nonfinite) = decode_record(&body[i * rec..i * rec + rec], schema);
        if row_nonfinite {
            nonfinite += 1;
        }
        rows.push(row);
    }

    let report = ParseReport {
        schema: schema.name,
        n_records: n,
        rec_size: rec,
        leftover_bytes: rem,
        layout_mismatch: rem != 0,
        nonfinite,
    };
    (rows, report)
}

/// Decode ONE fixed-size record into a `Row` + whether it carried a non-finite float. Shared by the whole-file
/// [`parse_frames`] and the streaming readers so they decode IDENTICALLY — `streaming-decode ≡ whole-file-decode`
/// by construction (the equivalence is asserted in `tests/streaming.rs`).
fn decode_record(rec_bytes: &[u8], schema: &Schema) -> (Row, bool) {
    let mut off = 0usize;
    let mut row: Row = Vec::with_capacity(schema.fields.len());
    let mut nonfinite = false;
    for &(name, ft) in schema.fields {
        match ft {
            FieldType::U64 => {
                let v = u64::from_le_bytes(rec_bytes[off..off + 8].try_into().unwrap());
                off += 8;
                row.push((name, Field::U64(v)));
            }
            FieldType::F32 => {
                let v = f32::from_le_bytes(rec_bytes[off..off + 4].try_into().unwrap());
                off += 4;
                if !v.is_finite() {
                    nonfinite = true;
                }
                row.push((name, Field::F32(v)));
            }
            FieldType::F64 => {
                let v = f64::from_le_bytes(rec_bytes[off..off + 8].try_into().unwrap());
                off += 8;
                if !v.is_finite() {
                    nonfinite = true;
                }
                row.push((name, Field::F64(v)));
            }
            FieldType::U8 => {
                row.push((name, Field::U8(rec_bytes[off])));
                off += 1;
            }
            FieldType::Pad => {
                off += 1; // consumed, dropped
            }
        }
    }
    (row, nonfinite)
}

/// Stream-parse fixed records from any `Read`, invoking `on_record(&Row)` per record — **bounded input memory**
/// (one record buffer; the whole dump is never loaded). Returns the same [`ParseReport`] the whole-file
/// [`parse_frames`] would, including `layout_mismatch` on a trailing partial record. Synchronous and
/// deterministic — NO async runtime, NO extra dependency (`std::io::BufReader` only). `streaming ≡ whole-file`
/// on the decisions; the input may arrive in arbitrary fragments (a record is filled across reads).
pub fn stream_frames<R: Read, F: FnMut(&Row)>(
    reader: R,
    schema: &Schema,
    header_lines: usize,
    mut on_record: F,
) -> std::io::Result<ParseReport> {
    let mut r = BufReader::new(reader);
    // skip `header_lines` newline-terminated text lines
    let mut skipped = 0usize;
    let mut one = [0u8; 1];
    while skipped < header_lines {
        match r.read(&mut one)? {
            0 => break,
            _ => {
                if one[0] == b'\n' {
                    skipped += 1;
                }
            }
        }
    }

    let rec = schema.rec_size();
    let mut buf = vec![0u8; rec.max(1)];
    let mut n_records = 0usize;
    let mut nonfinite = 0usize;
    let mut leftover_bytes = 0usize;
    if rec > 0 {
        loop {
            let mut filled = 0usize;
            while filled < rec {
                match r.read(&mut buf[filled..rec])? {
                    0 => break,
                    k => filled += k,
                }
            }
            if filled == 0 {
                break; // clean EOF on a record boundary
            }
            if filled < rec {
                leftover_bytes = filled; // trailing partial record
                break;
            }
            let (row, rn) = decode_record(&buf[..rec], schema);
            if rn {
                nonfinite += 1;
            }
            n_records += 1;
            on_record(&row);
        }
    }

    Ok(ParseReport {
        schema: schema.name,
        n_records,
        rec_size: rec,
        leftover_bytes,
        layout_mismatch: leftover_bytes != 0,
        nonfinite,
    })
}

/// Convenience: stream into a `Vec<Row>` — still bounded *input* buffering (reads incrementally rather than
/// loading the whole file as bytes). Mirrors [`parse_frames`]'s return so it is a drop-in for large dumps.
pub fn read_frames_streaming<R: Read>(
    reader: R,
    schema: &Schema,
    header_lines: usize,
) -> std::io::Result<(Vec<Row>, ParseReport)> {
    let mut rows = Vec::new();
    let report = stream_frames(reader, schema, header_lines, |row| rows.push(row.clone()))?;
    Ok((rows, report))
}

// ---- lift: obligations the emitted telemetry CAN support, kernel-relative (Sub-Slice 1B) ----------

/// Default containment bound (`U_MAX_DEFAULT` in the Python adapter).
pub const U_MAX_DEFAULT: f64 = 100.0;

fn has_field(rows: &[Row], name: &str) -> bool {
    rows.first().map_or(false, |r| field(r, name).is_some())
}

/// §7 containment from the pre-reduced max energy + frame count — the obligation builder shared by the
/// whole-file `containment` and the streaming fold, so both produce a BYTE-IDENTICAL `ObligationResult`.
/// `max_energy = None` (no energy field seen) ⇒ NaN ⇒ VIOLATED, matching the whole-file path.
pub fn containment_from(max_energy: Option<f64>, n_frames: usize, u_max: f64) -> ObligationResult {
    let mx = max_energy.unwrap_or(f64::NAN);
    let status = if mx.is_finite() && mx < u_max {
        ObligationStatus::Bounded
    } else {
        ObligationStatus::Violated
    };
    ObligationResult::new(
        "DVSM-3-kernel",
        "‖Z‖ stays under U_MAX on the REAL emitted telemetry",
        status,
        format!("max(energy)={:.3} over {} real frames (bound={})", mx, n_frames, u_max),
        "boundedness for all inputs — only this dumped run; empirical-boundedness ≠ certified",
        "a real dump whose energy reaches U_MAX without GhostSnap recovery",
    )
}

/// §7 containment: ‖Z‖ (the `energy` channel) stays under `u_max` on the emitted run.
pub fn containment(rows: &[Row], u_max: f64) -> ObligationResult {
    let mut max_energy: Option<f64> = None;
    for r in rows {
        if let Some(v) = field(r, "energy").and_then(|f| f.as_f64()) {
            max_energy = Some(match max_energy {
                Some(m) => m.max(v),
                None => v,
            });
        }
    }
    containment_from(max_energy, rows.len(), u_max)
}

/// The UNDERDETERMINED "no replay hash in this schema" obligation — shared by `lift` and `lift_streaming`.
fn no_hash_obligation(schema: &Schema) -> ObligationResult {
    ObligationResult::new(
        "DVSM-6-kernel",
        "replay-hash parity on the real dump",
        ObligationStatus::Underdetermined,
        format!("the {} frame carries no hash field — cannot check replay parity from this schema", schema.name),
        "anything about reproducibility from a hashless dump",
        "switch to the ABI/run_profile dump (carries the FNV-1a hash)",
    )
}

/// §6 replay parity: two real dumps from the same seed share an identical replay-hash sequence.
pub fn replay_parity(rows_a: &[Row], rows_b: &[Row]) -> ObligationResult {
    let a: Vec<Option<u64>> = rows_a.iter().map(|r| field(r, "hash").and_then(|f| f.as_u64())).collect();
    let b: Vec<Option<u64>> = rows_b.iter().map(|r| field(r, "hash").and_then(|f| f.as_u64())).collect();
    let status = if !a.is_empty() && a == b { ObligationStatus::Closed } else { ObligationStatus::Violated };
    ObligationResult::new(
        "DVSM-6-kernel",
        "two real dumps from the same seed share an identical replay-hash sequence",
        status,
        format!("{} vs {} frames; hash sequences identical = {}", a.len(), b.len(), a == b),
        "CORRECTNESS or cross-precision parity — integrity ≠ truth; hash ≠ reality",
        "identical seed yielding divergent emitted hashes",
    )
}

/// Forbidden-coupling obligations a PUBLIC frame dump cannot support: `(id, needed-field, why)`. Neither `v`
/// nor `lambda_eff` is emitted, so both are always non-liftable from the public schemas. `undetected ≠ absent`.
pub const NON_LIFTABLE_NEEDS: &[(&str, &str, &str)] = &[
    ("DVSM-7 (Ω→V air-gap)", "v", "velocity V is not emitted in the public frame; the Ω→V air-gap needs it"),
    ("DVSM-4 (ν→λ air-gap)", "lambda_eff", "dissipation λ is not emitted in the public frame; the ν→λ air-gap needs it"),
];

/// The forbidden-coupling obligations the emitted fields cannot support, with the reason — `(id, why)`.
pub fn non_liftable(rows: &[Row]) -> Vec<(&'static str, &'static str)> {
    NON_LIFTABLE_NEEDS
        .iter()
        .filter(|(_id, need, _why)| !has_field(rows, need))
        .map(|(id, _need, why)| (*id, *why))
        .collect()
}

/// Lift every obligation the dump supports to a graded [`ObligationResult`], and return the non-liftable ones
/// with why. Mirrors the Python `lift()`: containment if `energy` present; replay-parity if `hash` present and
/// a second dump is supplied, else an UNDERDETERMINED no-hash obligation for a hashless schema.
pub fn lift(
    rows: &[Row],
    schema: &Schema,
    u_max: f64,
    rows_b: Option<&[Row]>,
) -> (Vec<ObligationResult>, Vec<(&'static str, &'static str)>) {
    let mut obs = Vec::new();
    if has_field(rows, "energy") {
        obs.push(containment(rows, u_max));
    }
    if has_field(rows, "hash") {
        if let Some(rb) = rows_b {
            obs.push(replay_parity(rows, rb));
        }
    } else if !rows.is_empty() {
        obs.push(no_hash_obligation(schema));
    }
    (obs, non_liftable(rows))
}

/// Streaming `lift` for a SINGLE dump — bounded memory (one row + a running max energy, independent of file
/// size). Produces the SAME obligations + non-liftable set as `lift(rows, schema, u_max, None)` would on the
/// fully-buffered rows (the obligation builders are shared, so the equivalence is by construction; it is also
/// asserted in `tests/gateway.rs`). `streaming ≡ whole-file`. (A two-dump `replay_parity` is intentionally not
/// offered here — it would need both hash sequences held in lockstep, breaking the O(record) bound.)
pub fn lift_streaming<R: Read>(
    reader: R,
    schema: &Schema,
    u_max: f64,
    header_lines: usize,
) -> std::io::Result<(Vec<ObligationResult>, Vec<(&'static str, &'static str)>, ParseReport)> {
    let mut max_energy: Option<f64> = None;
    let mut first_row: Option<Row> = None;
    let report = stream_frames(reader, schema, header_lines, |row| {
        if first_row.is_none() {
            first_row = Some(row.clone()); // hold exactly ONE row (O(record))
        }
        if let Some(v) = field(row, "energy").and_then(|f| f.as_f64()) {
            max_energy = Some(match max_energy {
                Some(m) => m.max(v),
                None => v,
            });
        }
    })?;

    let mut obs = Vec::new();
    let non_lift;
    match &first_row {
        Some(fr) => {
            let probe = std::slice::from_ref(fr); // 1-row slice ⇒ `has_field` / `non_liftable` see the schema
            if has_field(probe, "energy") {
                obs.push(containment_from(max_energy, report.n_records, u_max));
            }
            if !has_field(probe, "hash") {
                obs.push(no_hash_obligation(schema));
            }
            non_lift = non_liftable(probe);
        }
        None => {
            non_lift = non_liftable(&[]); // empty stream
        }
    }
    Ok((obs, non_lift, report))
}

